ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -8,12 +8,48 @@
 module Development.IDE.Types.Logger
   ( Priority(..)
   , Logger(..)
+  , Recorder(..)
   , logError, logWarning, logInfo, logDebug, logTelemetry
   , noLogging
+  , WithPriority(..)
+  , logWith
+  , cmap
+  , cmapIO
+  , cfilter
+  , withDefaultRecorder
+  , makeDefaultStderrRecorder
+  , priorityToHsLoggerPriority
+  , LoggingColumn(..)
+  , cmapWithPrio
   ) where
 
-import qualified Data.Text as T
-
+import           Control.Concurrent         (myThreadId)
+import           Control.Concurrent.Extra   (Lock, newLock, withLock)
+import           Control.Exception          (IOException)
+import           Control.Monad              (forM_, when, (>=>))
+import           Control.Monad.IO.Class     (MonadIO (liftIO))
+import           Data.Functor.Contravariant (Contravariant (contramap))
+import           Data.Maybe                 (fromMaybe)
+import           Data.Text                  (Text)
+import qualified Data.Text                  as T
+import qualified Data.Text                  as Text
+import qualified Data.Text.IO               as Text
+import           Data.Time                  (defaultTimeLocale, formatTime,
+                                             getCurrentTime)
+import           GHC.Stack                  (HasCallStack, withFrozenCallStack)
+import           Prettyprinter              (Doc, Pretty (pretty),
+                                             defaultLayoutOptions, layoutPretty,
+                                             vcat, (<+>))
+import           Prettyprinter.Render.Text  (renderStrict)
+import           System.IO                  (Handle, IOMode (AppendMode),
+                                             hClose, hFlush, hSetEncoding,
+                                             openFile, stderr, utf8)
+import qualified System.Log.Formatter       as HSL
+import qualified System.Log.Handler         as HSL
+import qualified System.Log.Handler.Simple  as HSL
+import qualified System.Log.Logger          as HsLogger
```

Comment:
This is a significant change in behaviour. Switching to HsLogger seems unrelated to the contravariant logging changes, and better left for another PR

Reply:
The only reason for the hslogger stuff is so we can still see them in the logs because hie-bios and a few other places in the code base haven't been converted to the contravariant logging system.

Reply:
I don't follow this. Are you saying that the old stderr logger implementation didn't work with the contravariant system?

Reply:
Although now that I look, it seems the only place in hls that uses hslogger is the Hlint plugin, but I'm pretty sure I need the hslogger stuff to get the hiebios output for log files. The reason to copypaste the hiebios stuff was so the outputs wouldn't interleave improperly when we write to the handle.

Reply:
But the old stderr logger was already using locking to prevent undesired interleaving. I'm not keen on adding a new dependency in ghcide, specially as @michaelpj  is planning to drop the hslogger dependency in `lsp`

Reply:
Oh OK didn't see the newer message.

Previously HLS used both hslogger, and a plain stderr logger depending on the entry point. If `ghcide` process was run it would use stderr logger, and otherwise I think it would overwrite the default stderr logger with the hslogger logger. 
Additionally `hie-bios` uses hslogger to write compilation logs to output. The hslogger was initialized by calling `setupLogger` from `lsp`. If I kept using the `setupLogger` from `lsp` then the logs would get interleaved weirdly because they don't have access to the `lock`, and so I decided to copy-paste that code with some modifications so we could still get any logs written to some handle using hslogger, but also not have them displayed weirdly.

Reply:
Maybe the confusion is thinking that we're using hslogger as the console/file backend. We're not... we're using just hslogger to setup a hslogger so we can pick up `hie-bios` logs to the right handle, and give it a lock so it doesn't do funny things with messages we write to the actual handle backend.

Reply:
So, if I understand your analysis right, the `hie-bios` logs were actually broken for `ghcide`, since `hie-bios` unconditionally uses `hslogger`, and the `ghcide` entry point wasn't setting it up? That would also be true for the logs from `lsp`, which currently use `hslogger` (although I'm getting rid of it).

I'm okay with putting this in there as a temporary measure until we can get rid of `hslogger` in our dependencies. We should just clearly explain why we're doing it.

Reply:
Yes, I'll add a better comment.